stablehlo/reference/Api.cpp

```diff
@@ -170,19 +170,82 @@ LogicalResult removeDynamism(ModuleOp module, func::FuncOp func,
   return success();
 }
 
+bool isQuantizedTypes(TypeRange types) {
+  return llvm::any_of(types, [](Type type) {
+    return isa<quant::QuantizedType>(getElementTypeOrSelf(type));
+  });
+}
+
+// Recursively checks if an operation or any of its nested operations use
+// quantized types.
+//
+// Args:
+//   op: The operation to check for quantized type usage.
+//
+// Returns:
+//   True if the operation or any nested operation uses quantized types,
+//   false otherwise.
+bool operationUsesQuantType(Operation &op) {
+  for (Region &region : op.getRegions()) {
+    for (Block &block : region) {
+      for (Operation &op : block) {
+        TypeRange operandTypes = op.getOperandTypes();
+        TypeRange resultTypes = op.getResultTypes();
+        if (isQuantizedTypes(operandTypes) || isQuantizedTypes(resultTypes) ||
+            operationUsesQuantType(op)) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+// Lowers quantization-related operations and types within a function to
+// primitive math operations.
+//
+// This function checks if a function uses quantized types in its inputs,
+// outputs, or internal operations. If so, it creates and runs a StableHLO
+// quantization lowering pipeline to transform those quantized constructs into
+// primitive math operations. If the lowering process fails, an error is
+// emitted.
+//
+// Args:
+//   module: The module containing the function `func`.
+//   func The function to lower quantized types/operations in.
+//
+// Returns:
+//   A `LogicalResult` indicating success or failure of the lowering process.
+LogicalResult lowerQuantization(ModuleOp module, func::FuncOp func) {
+  if (!(isQuantizedTypes(func.getFunctionType().getInputs()) ||
+        operationUsesQuantType(*func.getOperation()) ||
+        isQuantizedTypes(func.getFunctionType().getResults()))) {
```

Comment:
Is this a performance optimization, or are we expecting that running the quantization lowering on a program without quant ops may nonetheless modify the program? If so, should we try to fix that?

Reply:
>> are we expecting that running the quantization lowering on a program without quant ops may nonetheless modify the program
Yes, that is the case based on the choice of passes in the pipeline. Per https://github.com/openxla/stablehlo/issues/2390 we believe that the pass pipeline can be simplified so that non-quantized programs will not get affected in which case we can use the pipeline unconditionally.

For more info: https://github.com/openxla/stablehlo/pull/2388#discussion_r1633735620